### ISSUE TITLE
New languages for <supportedLanguages> tag

### ISSUE THREAD

**Issue by sjabberwocky** [Gamua/Adobe-Runtime-Support#444]
Modern apps and games are localized into many languages.
It would be great to add new languages to <supportedLanguages> tag in app descriptor.
Recently I have done a research of the languages that players use on their devices.

These languages would be great to add to Adobe AIR:

Code | Language
------------ | -------------
ar | Arabic
fa | Persian
hi | Hindi
hu | Hungarian
ro | Romanian
uk | Ukrainian

All of them have financial sense to support in our app.

As bonus it would be nice to have Belarusian (be).

**Comment by JanFW** [comment]
> As bonus it would be nice to have Belarusian (be).
It should be **BY**; **BE** is for Belgium

**Comment by sjabberwocky** [comment]
> It should be **BY**; **BE** is for Belgium

It's language code, not a country code.
You can see ISO 639-2 as reference for them.
https://www.loc.gov/standards/iso639-2/php/code_list.php

**Comment by JanFW** [comment]
Alright, thanks for your clarification and sorry for the inconvenience!

**Comment by ajwfrost** [comment]
Reviewing this ticket - we're looking to see how complex it would be to add in some new languages. There may be some platform-specific limitations but an initial test suggested that for desktop AIR apps, you can just put in any code you want to a localized application name, but there are validation checks against the `supportedLanguages` field...

**Comment by dampgnat** [comment]
Hi Andrew, I currently have an app that successfully supports 15 languages but I'm unable to add all of these to the supportedLanguages tag without the invalid value error.  Is this the only way for AIR to create the info.plist for iOS?  Or do you know of any workarounds?  All we require is for the languages to show on the storefront.

**Comment by ajwfrost** [comment]
@dampgnat sorry we had lost track of this one... Are you able to download the below version of ADT and check? From our code reviews from before, it looks like this should work fine as long as you ensure you have the right resources in place for the language codes that you enter...
https://transfer.harman.com/message/X33u6l0ycLmvcm8CtvCxZi

thanks

**Comment by dampgnat** [comment]
Thanks for the quick response!
I updated to 1.1.698. Downloaded and replaced the adt file and packaged successfully. But when I add "ar" to the list of supported languages I get:
```
unexpected failure: null
java.lang.NullPointerException
	at com.adobe.air.ADTOutputStream.addFileFromStream(ADTOutputStream.java)
	at com.adobe.air.ipa.IPAOutputStream.addFileFromResource(IPAOutputStream.java)
	at com.adobe.air.ipa.IPAOutputStream.addLocalizationResources(IPAOutputStream.java)
	at com.adobe.air.ipa.IPAOutputStream.addApplicationDescriptor(IPAOutputStream.java)
	at com.adobe.air.ApplicationPackager.addSpecialFiles(ApplicationPackager.java)
	at com.adobe.air.ApplicationPackager.createPackage(ApplicationPackager.java)
	at com.adobe.air.ipa.IPAPackager.createPackage(IPAPackager.java)
	at com.adobe.air.ADT.parseArgsAndGo(ADT.java)
	at com.adobe.air.ADT.run(ADT.java)
	at com.adobe.air.ADT.main(ADT.java)
```
Can you explain what you mean by "ensure you have the right resources in place"? Do I need to add something else?

**Comment by ajwfrost** [comment]
Interesting, thanks .. so we missed the fact that there are actually localized strings that are provided for all of the AIR "supported languages". These are actually embedded into the ADT tool itself which is why we hadn't spotted them.

I think we'll have to put in place a fallback mechanism, so that where the strings aren't found internally, we'll check externally and then fall back to the first language in the list of supported languages..

thanks

**Comment by dampgnat** [comment]
The app uses an SDK to get the player's locale, not the device. So I guess whether AIR supports these languages or not is irrelevant in these scenarios, the store just needs to interpret the plist.  Hadn't considered that you would be able to check externally, curious!

Thanks for your excellent ongoing support for AIR Andrew!

**Comment by ajwfrost** [comment]
This is actually in ADT when it packages up the IPA file. If you unzip an IPA you'll see all the localized string resources under folders such as "en.lproj" - if you don't specify supported languages, then it just puts in all the ones that AIR supports. If you specify some languages then only those ones are added. But if you specify a language for which we don't have a translation within the ADT tool, it was failing.
Now we've set it to fall back to (a) seeing if there is a file in the input files that matches the expected pattern (languageCode.lproj/com.adobe.air.localizable.strings, you can see what format it should be in from the existing files); or failing that, (b) taking the first language code that's in the list and using that.
So the list of strings would need to have a valid language first (or to provide a manual file) otherwise it will fail with the same error...

Please try with this version then:
https://transfer.harman.com/message/hBSsaCdtoSiMNe6osbnM1o

thanks

**Comment by dampgnat** [comment]
YES!! Works great. No packaging errors and all 15 languages now list in AppStoreConnect. Thank you!!

**Comment by ajwfrost** [comment]
Excellent .. will get that tidied up and put into the next release...
thanks
